### ISSUE TITLE
CMake or meson support / buildsystem improvement

### ISSUE THREAD

**Issue by Mindavi** [AFLplusplus/AFLplusplus#811]
**Is your feature request related to a problem? Please describe.**

Hey, I'm (trying to be) a maintainer for the aflplusplus package in the [nixpkgs](https://github.com/nixos/nixpkgs) repository. However, for every new release I feel like more patches are required due to the homegrown build system which is quite strongly tied to the [Filesystem Hierarchy Standard](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard).

Is there any way to convince you to move to a build system with (better) support for non-FHS environment? Would a pull request be accepted if I decide to put the effort into building something like that (and of course make it all work as before)? Would there be a preference on which build system to use? Since you try to support quite a wide range of systems (as far as I remember), I could imagine that not every build system would be feasible.

E.g. see https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/security/aflplusplus/default.nix#L37-L51 for an idea of the patching required. Aflplusplus is also packages in the AUR (albeit with less support for all the plugins, I guess), where the build recipe is a lot simpler: https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=aflplusplus.

**Describe the solution you'd like**

I'd like it if a build system was chosen / used which doesn't rely heavily on a FHS-style OS / filesystem, to make packaging for NixOS / nixpkgs easier. Or at least a clear no, we will not try to support / do that. If it's a yes, I wouldn't mind taking a look into converting the project over, but I don't want to put in the time if a patch will never be accepted anyways.

**Describe alternatives you've considered**

Keep going through the pain of patching everything (I can't really keep up, nixpkgs is still on 2.65c because it's so much work to figure out what requires patching) or drop the support for aflplusplus from nixpkgs.

**Additional context**

I understand aflplusplus is quite a fast moving target, and would totally understand if this is unfeasible (either due to limited system support from other build systems (I guess cmake is quite widely supported?) or due to other reasons). But I really like it and would like to see it packaged properly for people willing to do fuzzing without having to figure out how to build aflplusplus.

Thanks for considering.

**Comment by vanhauser-thc** [comment]
We can only support one build system. it does not make sense to have two.
I personally would be willing to move to a different build system - if it supports all platforms where afl++ currently runs on ... from Solaris to Dragonfly to .... and old systems like Ubuntu 14.04.
Meson and ninja are too specific, new, and meson is python which I do not like as a dependency.
The only alternative I see is cmake (which I really do not like at all, but OK), but I do not know if it runs on all these platforms.

I am also not sure that cmake would really solve your issues? you still have to set CC/flags/... etc. so I am not sure what that actually would help (even after looking at your links). I have never worked on nixos, so likely I just lack the experience to understand your pain :)

So if cmake runs on all systems we support, and somebody else (e.g. you) are doing that, and the goal would be to replace all (!) makefiles, I would be ok. not sure how about the other team members think about that.

**Comment by domenukk** [comment]
Personally, I *love* the usability of meson (and don't like cmake too much), and the dependency on python vs on cmake makes no difference. Plus in fact we already have internal python scripts in the repo.
I do agree it would have to run on all systems then, which could be a massive pain (an may not be possible with meson yet?).

**Comment by hexcoder-** [comment]
I would like to understand the real issues there are in more detail before comtemplating a new build system (as that would mean more dependencies).
One weak point currently is: no parallel builds of afl++ source-only target.

**Comment by vanhauser-thc** [comment]
@Mindavi so what are you thoughts so far on this?

**Comment by Mindavi** [comment]
I think improving the current Makefile is the easiest solution. I'm going to attempt packaging again and will report back everything I run into, I think that's the easiest way.

I'm gonna close this and open specific issues when I run into them. Since there have been a couple of new releases for AFL++ since I did the packaging I think it would be best if I just report the individual issues I run into as I go.

I've also learned some things about make in the last couple of days which might come in handy (overriding variables and those kinds of things).

I think I was a little bit too eager opening this, sorry for that.
